File: Leetcode/Tree/226.invert-binary-tree.java
/*
Invert a binary tree.

Example:

Input:

     4
   /   \
  2     7
 / \   / \
1   3 6   9

Output:

     4
   /   \
  7     2
 / \   / \
9   6 3   1

*******************************************************************************Solution**********************************************************************/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution 
{
    //DFS based solution - preorder based
    public TreeNode invertTree(TreeNode root) 
    {
        if(root == null)
            return null;
        
        invertUtil(root);
        return root;
    }
    
    void invertUtil(TreeNode root)
    {
        //swapping the left and right subtree of current root node
        TreeNode temp = root.left;
        root.left = root.right;
        root.right = temp;
        
        invertTree(root.left);
        invertTree(root.right);
    }
    
    //BFS solution
//     public TreeNode invertTree(TreeNode root) 
//     {
//         Queue<TreeNode> q = new LinkedList<>();
//         if(root != null)
//             q.add(root);
        
//         while(!q.isEmpty())
//         {
//             TreeNode removed = q.poll();
            
//             //swapping the left and right subtree of current removed node
//             TreeNode temp = removed.left;
//             removed.left = removed.right;
//             removed.right = temp;
            
//             if(removed.left != null)
//                 q.add(removed.left);
            
//             if(removed.right != null)
//                 q.add(removed.right);
//         }
        
//         return root;
//     }    
}

File: DSA-JavaScript/Tree/226.invert-binary-tree.js
/*

Given the root of a binary tree, invert the tree, and return its root.


Example 1:

Input: root = [4,2,7,1,3,6,9]
Output: [4,7,2,9,6,3,1]

Example 2:

Input: root = [2,1,3]
Output: [2,3,1]

Example 3:

Input: root = []
Output: []
 

Constraints:

The number of nodes in the tree is in the range [0, 100].
-100 <= Node.val <= 100


**************************************************************Solution*****************************************************/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
const invertTree = function (root) {
  if (root === null) {
    return root;
  }

  const invertedLeft = invertTree(root.left);
  const invertedRight = invertTree(root.right);

  root.left = invertedRight;
  root.right = invertedLeft;
  return root;
};
